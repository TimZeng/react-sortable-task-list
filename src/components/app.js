import React, { Component } from 'react';

import SortableListComponent from './sortableList';

export default class App extends Component {
  render() {
    return (
      <div className='container'>
        <SortableListComponent />
      </div>
    );
  }
};