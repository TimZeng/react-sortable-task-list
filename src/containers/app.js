import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { arrayMove } from 'react-sortable-hoc';

import SortableListComponent from '../components/sortableList';
import { Button, InputGroup, Alert } from '../components/reusableComponents';
import { fetchTasks, saveTasks, updateGlobal, resetAlert } from '../actions/index';

class App extends Component {
  state = {
    tasks: [],
    newTask: ''
  };

  componentWillMount() {
    this.props.fetchTasks();
  }

  componentWillReceiveProps({ tasks: nextTasks }) {
    const { tasks: curTasks } = this.props;
    if (nextTasks !== curTasks) {
      this.setState({ tasks: nextTasks });
      this.props.updateGlobal({ addingTask: false, changeMade: false });
    }
  }

  renderAlert() {
    const { alert, resetAlert } = this.props;
    return (
      <Alert
        alert={ alert }
        onClose={ resetAlert }
      />
    )
  }

  renderButtons() {
    const { global, updateGlobal } = this.props;
    return (
      <div style={styles.headerStyle}>
        <span style={styles.titleStyle}>Tasks</span>
        <div>
          <Button
            text='Add Task'
            divStyle={{backgroundColor:'#8d9db0',color:'#fff',fontSize:'75%'}}
            disabled={global.addingTask}
            // onClick={() => updateGlobal({ addingTask: true })}
            onClick={this.addNewTask}
          />
          <Button
            text='Save'
            divStyle={{backgroundColor:'#78da9f',color:'#fff',fontSize:'75%',marginLeft:'8px'}}
            disabled={!global.changeMade}
            onClick={() => this.props.saveTasks(this.state.tasks)}
          />
        </div>
      </div>
    )
  }

  updateNewTask = newTask => {
    this.setState({ newTask });
  }

  // addNewTask = () => {
  //   const { tasks, newTask } = this.state;
  //   tasks.unshift(newTask);
  //   this.setState({ tasks, newTask: '' });
  //   this.props.updateGlobal({ addingTask: false, changeMade: true });
  // }

  addNewTask = () => {
    const { tasks } = this.state;
    tasks.unshift(null);
    this.setState({ tasks });
    // this.props.updateGlobal({ addingTask: false, changeMade: true });
  }

  removeTask = index => {
    const { tasks } = this.state;
    tasks.splice(index, 1);
    this.setState({ tasks });
    this.props.updateGlobal({ changeMade: true });
  }

  renderInputGroup() {
    const { global } = this.props;
    if ( !global.addingTask ) {
      return null;
    }

    const { newTask } = this.state;
    return (
      <InputGroup
        actionText='Add'
        value={newTask}
        onChange={this.updateNewTask}
        onClick={this.addNewTask}
      />
    )
  }

  onSortEnd = ({oldIndex, newIndex}) => {
    this.setState({
      tasks: arrayMove(this.state.tasks, oldIndex, newIndex),
    });
    this.props.updateGlobal({ changeMade: true });
  }

  renderSortableList() {
    const { tasks } = this.state;
    return (
      <SortableListComponent
        items={tasks}
        onSortEnd={this.onSortEnd}
        onDelete={this.removeTask}
      />
    )
  }

  render() {
    return (
      <div>
        <header>
          <h5>Sortable Task List - Tim Zeng - ReactJS & Redux</h5>
        </header>


        <div className='container'>
          { this.renderAlert() }
          <div className='content'>
            { this.renderButtons() }
            { this.renderInputGroup() }
            { this.renderSortableList() }
          </div>
        </div>
      </div>
    );
  }
};

const styles = {
  headerStyle: {
    display: 'flex',
    justifyContent: 'space-between',
    marginBottom: '30px'
  },

  titleStyle: {
    fontSize: '150%',
    fontWeight: '700',
    lineHeight: '38px'
  }
};

const mapStateToProps = ({ tasks, global, alert }) => ({ tasks, global, alert });

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchTasks, saveTasks, updateGlobal, resetAlert
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(App);
