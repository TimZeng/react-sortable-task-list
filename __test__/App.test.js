import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-15';
Enzyme.configure({ adapter: new EnzymeAdapter() });
import toJson from 'enzyme-to-json';

import { App } from '../src/containers/app';

describe('App component should render dynamically', () => {
  let wrapper, taskLen, alertCount, newTaskCount
  const fetchTasks = () => {};
  const global = { changeMade: false, tasksFetched: true };
  const newTask = 'new task for testing';
  const tasks = ['task1','task2','task3'];
  const successAlert = {status: 200, message: 'success'};

  it('should render newTasks when it exist', () => {
    const props = {fetchTasks, global, newTask, alert: {}, tasks };
    const component = shallow(<App {...props} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
    expect(component.find('input').length).toBe(1);
    expect(component.find('input').props().value).toEqual(newTask);
  })

  it('should NOT render newTasks when value is null', () => {
    const props = {fetchTasks, global, newTask: null, alert: {}, tasks };
    const component = shallow(<App {...props} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
    expect(component.find('input').length).toBe(0);
  })

  it('should render alert when it exist', () => {
    const props = {fetchTasks, global, newTask: null, alert: successAlert, tasks };
    const component = shallow(<App {...props} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  })

  it('should NOT render list item when tasks is empty', () => {
    const props = {fetchTasks, global, newTask: null, alert: {}, tasks:[] };
    const component = shallow(<App {...props} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  })

  it('should enable saving when list fetched and change made', () => {
    const props = {fetchTasks, global: { changeMade: true, tasksFetched: true }, newTask: null, alert: {}, tasks:[] };
    const component = shallow(<App {...props} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  })

});


