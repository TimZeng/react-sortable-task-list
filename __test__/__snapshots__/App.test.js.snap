// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render dynamically should NOT render list item when tasks is empty 1`] = `
<div>
  <header>
    <h5>
      Sortable Task List
    </h5>
  </header>
  <div
    className="container"
  >
    <Alert
      alert={Object {}}
    />
    <div
      className="content"
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
            "marginBottom": "30px",
          }
        }
      >
        <span
          style={
            Object {
              "fontSize": "150%",
              "fontWeight": "700",
              "lineHeight": "38px",
            }
          }
        >
          Tasks
        </span>
        <div>
          <Button
            disabled={false}
            divStyle={
              Object {
                "backgroundColor": "#8d9db0",
                "color": "#fff",
                "fontSize": "75%",
              }
            }
            onClick={[Function]}
            text="Add Task"
          />
          <Button
            disabled={true}
            divStyle={
              Object {
                "backgroundColor": "#78da9f",
                "color": "#fff",
                "fontSize": "75%",
                "marginLeft": "8px",
              }
            }
            onClick={[Function]}
            text="Save"
          />
        </div>
      </div>
      <SortableListComponent
        items={Array []}
        onDelete={[Function]}
        onSortEnd={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`App component should render dynamically should NOT render newTasks when value is null 1`] = `
<div>
  <header>
    <h5>
      Sortable Task List
    </h5>
  </header>
  <div
    className="container"
  >
    <Alert
      alert={Object {}}
    />
    <div
      className="content"
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
            "marginBottom": "30px",
          }
        }
      >
        <span
          style={
            Object {
              "fontSize": "150%",
              "fontWeight": "700",
              "lineHeight": "38px",
            }
          }
        >
          Tasks
        </span>
        <div>
          <Button
            disabled={false}
            divStyle={
              Object {
                "backgroundColor": "#8d9db0",
                "color": "#fff",
                "fontSize": "75%",
              }
            }
            onClick={[Function]}
            text="Add Task"
          />
          <Button
            disabled={true}
            divStyle={
              Object {
                "backgroundColor": "#78da9f",
                "color": "#fff",
                "fontSize": "75%",
                "marginLeft": "8px",
              }
            }
            onClick={[Function]}
            text="Save"
          />
        </div>
      </div>
      <SortableListComponent
        items={
          Array [
            "task1",
            "task2",
            "task3",
          ]
        }
        onDelete={[Function]}
        onSortEnd={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`App component should render dynamically should enable saving when list fetched and change made 1`] = `
<div>
  <header>
    <h5>
      Sortable Task List
    </h5>
  </header>
  <div
    className="container"
  >
    <Alert
      alert={Object {}}
    />
    <div
      className="content"
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
            "marginBottom": "30px",
          }
        }
      >
        <span
          style={
            Object {
              "fontSize": "150%",
              "fontWeight": "700",
              "lineHeight": "38px",
            }
          }
        >
          Tasks
        </span>
        <div>
          <Button
            disabled={false}
            divStyle={
              Object {
                "backgroundColor": "#8d9db0",
                "color": "#fff",
                "fontSize": "75%",
              }
            }
            onClick={[Function]}
            text="Add Task"
          />
          <Button
            disabled={true}
            divStyle={
              Object {
                "backgroundColor": "#78da9f",
                "color": "#fff",
                "fontSize": "75%",
                "marginLeft": "8px",
              }
            }
            onClick={[Function]}
            text="Save"
          />
        </div>
      </div>
      <SortableListComponent
        items={Array []}
        onDelete={[Function]}
        onSortEnd={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`App component should render dynamically should render alert when it exist 1`] = `
<div>
  <header>
    <h5>
      Sortable Task List
    </h5>
  </header>
  <div
    className="container"
  >
    <Alert
      alert={
        Object {
          "message": "success",
          "status": 200,
        }
      }
    />
    <div
      className="content"
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
            "marginBottom": "30px",
          }
        }
      >
        <span
          style={
            Object {
              "fontSize": "150%",
              "fontWeight": "700",
              "lineHeight": "38px",
            }
          }
        >
          Tasks
        </span>
        <div>
          <Button
            disabled={false}
            divStyle={
              Object {
                "backgroundColor": "#8d9db0",
                "color": "#fff",
                "fontSize": "75%",
              }
            }
            onClick={[Function]}
            text="Add Task"
          />
          <Button
            disabled={true}
            divStyle={
              Object {
                "backgroundColor": "#78da9f",
                "color": "#fff",
                "fontSize": "75%",
                "marginLeft": "8px",
              }
            }
            onClick={[Function]}
            text="Save"
          />
        </div>
      </div>
      <SortableListComponent
        items={
          Array [
            "task1",
            "task2",
            "task3",
          ]
        }
        onDelete={[Function]}
        onSortEnd={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`App component should render dynamically should render newTasks when it exist 1`] = `
<div>
  <header>
    <h5>
      Sortable Task List
    </h5>
  </header>
  <div
    className="container"
  >
    <Alert
      alert={Object {}}
    />
    <div
      className="content"
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
            "marginBottom": "30px",
          }
        }
      >
        <span
          style={
            Object {
              "fontSize": "150%",
              "fontWeight": "700",
              "lineHeight": "38px",
            }
          }
        >
          Tasks
        </span>
        <div>
          <Button
            disabled={false}
            divStyle={
              Object {
                "backgroundColor": "#8d9db0",
                "color": "#fff",
                "fontSize": "75%",
              }
            }
            onClick={[Function]}
            text="Add Task"
          />
          <Button
            disabled={true}
            divStyle={
              Object {
                "backgroundColor": "#78da9f",
                "color": "#fff",
                "fontSize": "75%",
                "marginLeft": "8px",
              }
            }
            onClick={[Function]}
            text="Save"
          />
        </div>
      </div>
      <div
        className="light-text list-input"
      >
        <input
          autoFocus={true}
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="New task"
          value="new task for testing"
        />
        <i
          className="button fa fa-trash-o"
          onClick={[Function]}
          style={
            Object {
              "lineHeight": "24px",
            }
          }
        />
      </div>
      <SortableListComponent
        items={
          Array [
            "task1",
            "task2",
            "task3",
          ]
        }
        onDelete={[Function]}
        onSortEnd={[Function]}
      />
    </div>
  </div>
</div>
`;
